A couple of improvements to the framework

vrurg opened the pull request:
I'm playing with writing own simplistic `Term::UI` framework and decided to base it on your work. But a couple of changes was needed to make this possible. So, here is the result.

Besides, I have implemented basic support for terminal size change by allowing on-the-fly resetup of `Terminal::Print` instance. It's up to the user application though to react on `SIGWINCH` and initiate the reconfiguration.

japhb commented on lib/Terminal/Print/Grid.pm6:
```diff
@@ -8,7 +8,7 @@ unit monitor Terminal::Print::Grid;
 my class Cell {
     use Terminal::ANSIColor; # lexical imports FTW
-    has $.char is required;
+    has Str:D $.char is required where *.chars == 1;
```
Have you checked the performance of this change?  Cell is pretty bare bones because it needs to be really fast ....

vrurg replied:
There is an impact, no doubt. But it's a trade-off. And the impact itself is rather insignificant, based on `mandelbrot-pixels.p6` run:

- withoug constrain:
   ```
   16 zooms of 11676 pixels each in 1.875 seconds = 99635 pixels/second
   ```

- with constrain:
   ```
   16 zooms of 11676 pixels each in 1.905 seconds = 98047 pixels/second
   ```

The point of adding this was to prevent weird behavior when I was passing a mutli-char string instead of a single char. That was before I got better understanding of how things works inside. For now I wouldn't need this, but for somebody it could spare some time on fixing a bug.

The reason of having it here is because it's the single point of failure. Perhaps I overreacted and it'd be better to constrain parameters of `change-cell` and `print-cell`. Though it would still be a good thing to catch possible bugs.

vrurg replied:
Updated. Sometimes the code with constrain turned on displays even better results than the code without constrain. Either there is not so many writes directly into a cell or Raku optimization kicks in.

japhb commented on lib/Terminal/Print/Grid.pm6:
```diff
@@ -182,6 +182,8 @@ multi method change-cell($x, $y, %c) {
     @!grid[$y][$x] = Cell.new(|%c);
 }
 
+multi method change-cell($x, $y, *%c) { samewith($x, $y, %c) }
```
samewith is very slow.  This should probably be an explicit call (and the same with the other one).

vrurg replied:
This can just be removed as well. Just a convenience sugar for being able use plain list of pairs and not wrap it into `{ }`.

japhb commented on lib/Terminal/Print/Grid.pm6:
```diff
@@ -216,6 +218,8 @@ multi method print-cell($x, $y, %c) {
     self.print-cell($x, $y);
 }
 
+multi method print-cell($x, $y, *%c where *.elems) { samewith($x, $y, %c) }
```
A where clause can be expensive (as can the samewith).

vrurg replied:
Same as above. Without `where` it conflicts with `print-cell($x, $y)`. Because it's just for convenience, can be just removed.

japhb commented on lib/Terminal/Print/Grid.pm6:
```diff
@@ -44,12 +44,28 @@ method rows    { $!h }
 #| Instantiate a new (row-major) grid of size $w x $h
-method new($w, $h, :$move-cursor = move-cursor-template) {
-    my @grid = [ [ ' ' xx $w ] xx $h ];
+method new($w, $h, :@grid is copy, :$move-cursor = move-cursor-template) {
```
Where is the fast path?  You've converted the stripped-and-fast new to a slow path.

vrurg replied:
I don't think creating a new grid is a high-repetitive task which would make such a measurable impact on overall performance. Besides, the condition body wouldn't be executed is `@grid` is empty. Unfortunately, I don't have much time right now to bench `xx` vs. a loop, but it's unlikely that the difference would be that significant.

vrurg replied:
Here is interesting benchmark:

```perl6
my @p =
    start {
        my $st = now;
        for ^10000 {
            my @g = [ [ ' ' xx 100 ] xx 100 ];
        }
        my $et = now;
        note "xx + xx: ", $et - $st;
    },
    start {
        my $st = now;
        for ^10000 {
            my @g;
            for ^100 -> $gr {
                my @row = ' ' xx 100;
                if False {
                    note "IMPOSSIBLE";
                }
                @g[$gr] = @row;
            }
        }
        my $et = now;
        note "for + xx: ", $et - $st;
    };

await @p;
```

And here is its output:

```
for + xx: 12.6258859
xx + xx: 14.7326081
```

The outcome is pretty much consistent.

*NOTE* I do it in parallel because consequent execution on my laptop results in CPU heating and throttling. So, second tests might display biased results. `if` simulates when no `@grid` supplied.

japhb commented on lib/Terminal/Print.pm6:
```diff
@@ -198,22 +223,22 @@ class Terminal::Print {
-        die "No grid has been named $name" unless my $grid-index = %!grid-name-map{$name};
-        @!grids[$grid-index].grid;
+        die "No grid has been named $name" unless %!grid-name-map{$name}:exists;
+        @!grids[ %!grid-name-map{$name} ].grid;
```
I'm not sure I see the purpose of this patch block, except to do the hash lookup twice.

vrurg replied:
Because grid index can be 0 which is _False_ but valid value. `:exists` is cheap enough and the lookup itself is quick.

It could be `without my $grid-index = %!grid-name-map{$name}` which is more correct. But still, an existing key with undefined value is a bug.

japhb commented on lib/Terminal/Print.pm6:
```diff
@@ -198,22 +223,22 @@ class Terminal::Print {
-        die "No grid has been named $name" unless my $grid-index = %!grid-name-map{$name};
-        @!grids[$grid-index];
+        die "No grid has been named $name" unless %!grid-name-map{$name}:exists;
+        @!grids[ %!grid-name-map{$name} ];
```
Same comment as above -- this just means doing the hash lookup twice.

japhb commented on lib/Terminal/Print.pm6:
```diff
@@ -99,9 +94,39 @@ class Terminal::Print {
+    multi method root-widget(Terminal::Print::Widget $widget-type?) {
         my $grid = self.current-grid;
-        %!root-widget-map{$grid} ||= Terminal::Print::Widget.new-from-grid($grid);
+        %!root-widget-map{$grid} ||= $widget-type.WHAT.new-from-grid($grid);
```
Why not make constrain the parameter to :U and avoid the .WHAT indirection?

vrurg replied:
Left-over from earlier implementation before I made it a multi. Will fix.

japhb commented on lib/Terminal/Print.pm6:
```diff
@@ -76,20 +76,15 @@ class Terminal::Print {
-        push @!grids, $!current-grid;
+        unless PROCESS::<$TERMINAL>:exists {
+            PROCESS::<$TERMINAL> = self;
+        }
```
PROCESS::<$TERMINAL> //= self;

vrurg replied:
I wouldn't overwrite a symbol no matter what it contains. It might be like this on purpose and set by something else.

japhb commented on lib/Terminal/Print.pm6:
```diff
@@ -99,9 +94,39 @@ class Terminal::Print {
     }
 
-    method root-widget() {
+    method setup(:$reset = False) {
```
What about resize *without* deleting contents?

vrurg replied:
Not sure about preserving the content. It feels more correct for the content to be re-generated as the layout must change. This is a reason why I didn't move `SIGWINCH` processing in `Terminal::Print` and left it up to an app developer. As soon as it is known that terminal size changed he/she can take measures either to restore the old content or repaint from scratch.

japhb commented:
Generally speaking, a lot of these changes result in pessimization.  I'm certainly open to making a few things easier -- I've thought of creating a widget library of my own but just haven't had time, so the idea of e.g. allowing resizes makes sense to me -- but it's important to allow other users of the library (who want to make e.g. animations) to have fast paths.

vrurg commented:
Sorry, I don't get it what kind of pessimization do you mean? Besides, it looks like my changes don't break compatibility as tests are passing and examples are running.

vrurg commented:
I have comitted a few fixes. Seemingly I missed rpg-ui when tested examples, it was broken by the constrain on `char` in `Cell`. Implemented a quick workaround for it by making `set-span` accept same parameters, as `change-cell`. Otherwise it would require more time to rewrite parts of the example code which I don't have time for. But making `set-span` accept `string` key alongside with `char` would actually make it an acceptable fix.

ab5tract commented:
Hey @vrurg  and @japhb , I'm really happy to see this discussion. Sorry for joining late!

I think this PR has some real merit but it will take me until the weekend at least to give it a proper analysis.

In general I think resolving surprise issues and random discordia is a reasonable offset for small dips in performance. Especially since `Terminal::Print` keeps getting faster based on improvements to `raku` over any explicit attempts on our own :)

Also I think this patch might be a nice (near) end cap on the V1 API -- which might have too much sugar to support the real potential of its true bubbly personality anyway :)

ab5tract commented:
@vrurg all that I may ask is that you consider joining the 21st century and sparing a few extra characters on your `Term::UI` library to join us in `Terminal::*` ;) ;) ;)

vrurg commented:
@ab5tract Thanks for the good words!

A few notes. Since it all started, I have renamed `Term::UI` into [`Vikna`](https://github.com/vrurg/raku-Vikna). Mainly because `Term::UI` in Perl means different thing but also to have one-worded project name. Besides, the ideology of the project shifted significantly towards even-driven and parallelized paradigm. Also I have switched to own `Canvas` concept instead of using grids...

I mean, eventually, all I need of `Terminal::*` is `Terminal::Print::Commands` and `Terminal::ANSIColor`. So, this PR is basically orphaned, I don't think I would have time to do any changes in it any time soon. OTOH, any help with `Vikna` would be appreciated! :)
